api/src/main/java/org/eclipse/microprofile/concurrent/ManagedExecutorBuilder.java

```diff
@@ -44,9 +44,11 @@
      * instances.</p>
      *
      * @return new instance of <code>ManagedExecutor</code>.
-     * @throws IllegalArgumentException if the same thread context type is
-     *         present in, or inferred by, both the {@link #cleared} set
-     *         and the {@link #propagated} set.
+     * @throws IllegalArgumentException if a thread context type that is
```

Comment:
I think in this case having the `DefinitionException` would be more appropriate. There are no actual arguments in the method itself, hence `IllegalArgumentException` seems a bit off to me.

Reply:
We cannot use DefinitionException here because it would cause the builders to have a dependency on CDI and the whole reason the builders were added to the spec was to provide a way to use managed executors/thread context without CDI.  This was requested by Stéphane under https://github.com/eclipse/microprofile-sandbox/issues/18

That said, I'm certain open to a better exception class when the build method detects that an invalid combination of config settings were specified on the builder instance.  I'm sure this isn't the first time a spec with builders has encountered a pattern like this so maybe we can see if there is a precedent for it.

Reply:
After a bit of searching I found the following guidance in [Effective Java](https://books.google.com/books?id=ka2VUBqHiWkC&pg=PA15&lpg=PA15#v=onepage&q&f=false) written by Joshua Bloch: "_If any invariants are violated, the build method should throw an IllegalStateException. The exception's detail message should indicate which invariant is violated._"

I do also see where IllegalArgumentException is used in some existing specs, for example, in the [JsonbBuilder.build](https://static.javadoc.io/javax.json.bind/javax.json.bind-api/1.0/javax/json/bind/JsonbBuilder.html#build--)

It doesn't really concern me which exception type is used here so much as it does clearly defining what the error conditions are.  I can switch it over to IllegalStateException to be consistent with _Effective Java_.